### PR TITLE
Add Enhanced Data Analytics

### PR DESCRIPTION
**Pull request by saadm2**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by VenturaDelMonte** [comment]
Thank you for your contribution. @jeyhunkarimov or I ll review this next week. Please bear with us.

**Comment by saadm2** [comment]
WeightedMomentsEstimator

**Comment by jeyhunkarimov** [comment]
@saadm2 can you please rebase?

**Comment by VenturaDelMonte** [comment]
@jeyhunkarimov @saadm2 I changed the target branch. No worries.

**Comment by saadm2** [comment]
I have changed my default branch to develop.

File: src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+  }
+  case object AdaptiveReservoirThreshold extends Parameter[Double] {
+    override val defaultValue: Option[Double] = Some(0.99)
+  }
+  case object Change extends Parameter[Int] {
+    override val defaultValue: Option[Int] = Some(5)
+  }
+
+  //=====================================Extra=================================================
+  def factorial(n: Int): Int = {   if (n == 0) 1 else n * factorial(n-1)  }
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala]: Please, use tail-recursive factorial implementation.

File: src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+                                        input: DataStream[T])
+      : DataStream[mutable.ArrayBuffer[T]] = {
+        val resultingParameters = instance.parameters ++ transformParameters
+        val statefulStream = FlinkSolmaUtils.ensureKeyedStream[T](input, resultingParameters.get(PartitioningOperation))
+        var k = resultingParameters(AdaptiveReservoirSize)
+        val nboCh=resultingParameters(AdaptiveReservoirNumberToChange)
+        val thresh=resultingParameters(AdaptiveReservoirThreshold)
+        val changedReservoirSize = resultingParameters(Change)
+        val gen = new XORShiftRandom()
+        implicit val typeInfo = TypeInformation.of(classOf[(Long,Long,Long,mutable.ArrayBuffer[T],mutable.ArrayBuffer[T], mutable.ArrayBuffer[T])])
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala]: Please, use createTypeInformation macro as done in other parts of SOLMA.

File: src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        implicit val retInfo = TypeInformation.of(classOf[mutable.ArrayBuffer[T]])
+        var j : Int = 0
+        var m : Int = 0
+        var x : Int = 0
+        var remainingSize = 0
+        statefulStream.flatMapWithState((in, state: Option[(Long,Long,Long,mutable.ArrayBuffer[T],mutable.ArrayBuffer[T],mutable.ArrayBuffer[T])]) => {
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala]: Too big tuples should be avoided. Perhaps, you may consider to use a case class.

File: src/main/scala/eu/proteus/solma/sampling/WeightedReservoirSampling.scala [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      override def transformDataStream(
+                                        instance: WeightedReservoirSampling,
+                                        transformParameters: ParameterMap,
+                                        input: DataStream[T])
+      : DataStream[mutable.ArrayBuffer[T]]= {
+        val resultingParameters = instance.parameters ++ transformParameters
+        val statefulStream = FlinkSolmaUtils.ensureKeyedStream[T](input,resultingParameters.get(PartitioningOperation))
+        val k = resultingParameters(WeightedReservoirSize)
+        val gen = new XORShiftRandom()
+        implicit val typeInfo = TypeInformation.of(classOf[(Long,mutable.ArrayBuffer[Double], mutable.ArrayBuffer[T])])
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sampling/WeightedReservoirSampling.scala]: same as above.

**Comment by VenturaDelMonte** [comment]
Moreover, it would be nice to have a message describing what a PR adds in order to better understand the submitted code.

**Comment by VenturaDelMonte** [comment]
Also, please rebase your current PR on the actual develop branch as we enabled code-style checks to improve code maintainability.

File: src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  }
+  case object AdaptiveReservoirThreshold extends Parameter[Double] {
+    override val defaultValue: Option[Double] = Some(0.99)
+  }
+  case object Change extends Parameter[Int] {
+    override val defaultValue: Option[Int] = Some(5)
+  }
+
+  //=====================================Extra=================================================
+  def factorial(n: Int): Int = {   if (n == 0) 1 else n * factorial(n-1)  }
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala]: please convert this method as tail-recursive (and use `@tailrec` annotation)

File: src/test/scala/eu/proteus/solma/moments/WeightedMomentsEstimatorITSuite.scala [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+       }
+     }
+   }
+
+}
+
+case class WeightedSample (
+                    s: IndexedSeq[Int],
+                    w:DenseVector,
+                    d: DenseVector) extends StreamEvent2 {
```
Comment (VenturaDelMonte) [inline src/test/scala/eu/proteus/solma/moments/WeightedMomentsEstimatorITSuite.scala]: I do not see this class in the unified view of commits, meaning this class does not exist, hence, you did not compile your changes.  Please, always make sure to commit tested code.

File: src/main/scala/eu/proteus/solma/events/StreamEvent.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  */
+trait StreamEvent extends Serializable
+{
+  def slice: IndexedSeq[Int]
+  def data: Vector
+}
+
+trait WeightedStreamEvent extends StreamEvent
+{
+  def slice: IndexedSeq[Int]
```
Comment (saadm2) [inline src/main/scala/eu/proteus/solma/events/StreamEvent.scala]: Sorry, my mistake!

**Comment by saadm2** [comment]
I have carried out the requested updates. 

Just a small issue, i do not understand what do you mean by "we would like to see a rebase with upstream/develop for the next updates". Frankly, i have very little experience with github. Should I open a new pull request to upstream/develop ?

Many Thanks
Saad

**Comment by jeyhunkarimov** [comment]
@saadm2 Thanks for your response. It means that we would like to see a rebased commit. You don't need to open a new pull request. You can take a look to `git rebase` documentation from the web. Moreover, can you please point us where (code, commit ...) you handled the requested updates?

**Comment by VenturaDelMonte** [comment]
@saadm2 I am gonna try to help you. Assuming you have your current develop branch selected, the conflict in StreamEvent solved, and a remote (e.g., upstream) pointing to *this* repository. Please, execute `git fetch upstream` and `git rebase upstream/develop`. This will forward your changes on top of the current develop branch. Then, assuming no errors have happened, please execute `mvn verify` and solve the eventual errors accordingly. Let us know if you need further help, please.

**Pull request merged**